core/trino-main/src/main/java/io/trino/sql/planner/LocalExecutionPlanner.java

```diff
@@ -2144,8 +2143,8 @@ public PhysicalOperation visitIndexSource(IndexSourceNode node, LocalExecutionPl
             };
 
             // Declare the input and output schemas for the index and acquire the actual Index
-            List<ColumnHandle> lookupSchema = Lists.transform(lookupSymbolSchema, forMap(node.getAssignments()));
-            List<ColumnHandle> outputSchema = Lists.transform(node.getOutputSymbols(), forMap(node.getAssignments()));
```

Comment:
This is OK change here, but generally `Lists.transform` shouldn't be forbidden. 
For randomly accessed list, `Lists.transform`'s laziness (transformation upon access) may be way way cheaper.
Not to mention that `Lists.transform` is conceptually a _view_, not a copy.

Reply:
I honestly not a big fan of the `transforms` interface. It could be error prone, and we've had issues with this in the past. When such a list is passed somewhere it isn't obvious that operations which intuitively should be extremely cheap (such as `get(index)`) could actually be rather expensive. For lazy collections if feels like it is better when the "lazyness" is expressed at the interface level.

Reply:
>  For lazy collections if feels like it is better when the "lazyness" is expressed at the interface level.

is there a LazyList interface that i can pass where a List is expected?

> When such a list is passed somewhere it isn't obvious that operations which intuitively should be extremely cheap (such as get(index)) could actually be rather expensive

Yes, it's creator's responsibility to ensure they are creating lazy list purposefully.
It's certainly NOT the typical case, and I agree that `Stream`, `map`, `collect` is the recommended pattern.

Reply:
> Yes, it's creator's responsibility to ensure they are creating lazy list purposefully.

I agree with that. But from the code that does manipulations on a list it might not necessarily be obvious.

Reply:
Well, if i write a function that just does a transform on a list, returning a result, it must be defined, whether this is a transformed copy or a transforming view. It's NOT "whichever is more convenient to write", or "whichever happens to be preferred code style of the function author"

In the absence of documentation, immutable transformed copy should be assumed desired.

Reply:
Let's say there are 3 functions:


```
/**
*  This method is well documented and returns a lazy list, computations are expensive and it retains way more memory than needed.
*/
List<Integer> getSomeUsefulListOfIntegersByAuthorA(){
   List<SomeVeryHeavyDatabaseObject> list = fetchFromDatabase();
   return transform(list, SomeVeryHeavyDatabaseObject::doSomeModeratelyExpensiveComputationToExtractAnInteger);
}

SomeEntitiy doSomethignUsefulByAuthorB(){
   List<Integer> integers = getSomeUsefulListOfIntegersByAuthorA();
   // I read the documentation and I know that the list is lazy, I also checked the implementation of the processInAUsefulWayAndReturnSomething and it is fine to pass a lazy list
   return processInAUsefulWayAndReturnSomething(integers);
}

SomeEntity processInAUsefulWayAndReturnSomethingByAuthorCModifiedByAuthorD(List<Integer> integers ){
  // old implementation
  // return db.fetch(integers);
  /
  / new implementation
  // Let's check preconditions:
  checkArgument(integers.stream().allMatch(integer -> integer > 0));
  checkArgument(integers.stream().allMatch(integer -> integer < 2));
  // And maybe cache it?
  return cache.get(integers).orElse(() -> db.fetch(integers));
}
```

Now after a seemingly "innocent" refactor there's a serious memory leak and a major increase in the cost of computation.

I agree that in theory there might be some use cases for `Lists.transform` (for example using it only locally within a scope of a single method) thus this ban comes with a cost. But do you think the cost of banning this API is higher than the cost of potential damage it may cause?